pandas/core/arrays/interval.py

```diff
@@ -210,22 +192,83 @@ def __new__(
             )
             closed = closed or infer_closed
 
-        closed = closed or "right"
-        left, right = _maybe_cast_inputs(left, right, copy, dtype)
-        combined = _get_combined_data(left, right)
-        result = cls._simple_new(combined, closed=closed)
-        if verify_integrity:
-            result._validate()
-        return result
+        return cls._simple_new(
+            left,
+            right,
+            closed,
+            copy=copy,
+            dtype=dtype,
+            verify_integrity=verify_integrity,
+        )
 
     @classmethod
-    def _simple_new(cls, data, closed="right"):
+    def _simple_new(
+        cls, left, right, closed=None, copy=False, dtype=None, verify_integrity=True
+    ):
         result = IntervalMixin.__new__(cls)
 
-        result._combined = data
-        result._left = data[:, 0]
-        result._right = data[:, 1]
+        closed = closed or "right"
+        left = ensure_index(left, copy=copy)
+        right = ensure_index(right, copy=copy)
+
+        if dtype is not None:
+            # GH 19262: dtype must be an IntervalDtype to override inferred
+            dtype = pandas_dtype(dtype)
+            if not is_interval_dtype(dtype):
+                msg = f"dtype must be an IntervalDtype, got {dtype}"
+                raise TypeError(msg)
+            elif dtype.subtype is not None:
+                left = left.astype(dtype.subtype)
+                right = right.astype(dtype.subtype)
+
+        # coerce dtypes to match if needed
```

Comment:
can you factor this out into a function (maybe_cast_inputs), this is crazy otherwise

Reply:
This is how it was before (I am just reverting / fixing conflicts). I prefer to keep this PR a clean revert, and do any other refactor as follow-up

Reply:
i c ok.